subsys: testsuite: coverage: support resetting counters

srv-meta opened the pull request:
This continues the work started by Joshua in https://github.com/zephyrproject-rtos/zephyr/pull/65842

Ability to reset gcov counters allows better isolation of tests
coverage. Specifically it allows to:
1. Not include counters for any code that was executed prior running the
   test, which includes all the code executed during the board boot.
2. Run multiple tests without firmware reload by resetting counters
   between each.


I also undo renaming of gcov_populate_buffer to gcvo_to_gcda from bc8b7dd6dd36e0ae6b256353bf967c3568b907a7 as it was actually a bug. The function was only renamed in .c, but not in the corresponding .h. I opted for returning it to original name to avoid breaking the public contract from the header.

srv-meta committed:
testsuite: coveerage: support counters reset

Ability to reset gcov counters allows better isolation of tests
coverage. Specifically it allows to:
1. Not include counters for any code that was executed prior running the
   test, which includes all the code executed during the board boot.
2. Run multiple tests without firmware reload by resetting counters
   between each.

Signed-off-by: Roman Studenikin <srv@meta.com>

ycsin commented:
cc @nordic-piks

ycsin commented on subsys/testsuite/coverage/coverage.c:
```diff
@@ -297,7 +324,7 @@ void gcov_coverage_dump(void)
 {
 	/* Locking someway before getting this is recommended. */
-	return gcov_info_head;
+    return gcov_info_head;
```
[editorconfig](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig) will guarantee that the right indentation is used

ycsin reviewed (commented):
Apart from the indentation issue, commit needs a `Signed-off-by: ..`, otherwise LGTM

jgl-meta commented on subsys/testsuite/coverage/coverage.c:
```diff
@@ -232,7 +232,34 @@ size_t gcov_to_gcda(uint8_t *buffer, struct gcov_info *info)
+	uint32_t iter_counts;
+	uint32_t iter_counter_values;
+
+	for (iter_functions = 0U;
```
I think this traversal code is copy paste? Maybe extract it into it's own function?

srv-meta replied:
I am not sure it will make the code more readable. And I don't see any other place where it could be reused

cfriedt replied:
That would be kind of useful, even in terms of public API. The major benefit I think would be that external code could iterate over coverage data while keeping it in a single, private, location. This would enable coverage backends.

srv-meta replied:
There are 3 nested loops here and there are actions that are needed to be done on each level. Do you propose to define a traversal method that contains these 3 loops and accepts 3 callback functions? And each function will receive a context consisting of multiple variables?
I really don't see how to make it in a simple way that would improve things. Please suggest the exact api you are proposing

jgl-meta replied:
That is kind of an interesting proposal. I hadn't thought about it with three callbacks, each one being applied to each level. 

What I was thinking is that mostly the "outer" two for loops are just finding out how to walk the counters. I was more thinking about a single function to operate on the specific counters themselves, which are really what the point of the code is in all of the places it is called. So the idea is, to provide something like a for each functionality for each counter.

jgl-meta replied:
I think a seperate PR is acceptable or we could talk offline. A test would still be nice to see, but in the name of getting things moving I will accept as I don't see any obvious issues with the code.

jgl-meta commented:
A test would be nice to see.

nordic-piks commented:
@srv-meta Thanks for fixing gcov_populate_buffer to gcvo_to_gcda.

I have one question, when (at which conditions) `gcov_reset_counts` will be called?
Will it be done automatically and unconditionally? or rather from external trigger (like debugger)?
What you wrote here:
> Not include counters for any code that was executed prior running the
> test, which includes all the code executed during the board boot.

has downside, especially for testing at HW, as during the board boot there are eg. drivers and peripherals initialized
and IMHO we want to have coverage from that part.

srv-meta commented:
@nordic-piks 
This is just a toolkit that one can opt to use or not to use. I don't think any specific pattern should be enforced. One can opt to measuring coverage of the boot sequence as a separate test or not measure it at all. Someone else might not need to use resets at all as they don't have a requirement of measuring coverage of each individual test so precisely.

ycsin reviewed (dismissed):
LGTM

nashif commented:
would like to see this used in the tree, i.e. add an kconfig option that would allow us to reset counters before running a test.

srv-meta committed:
testsuite: coverage: Add CONFIG_ZTEST_COVERAGE_RESET_BEFORE_TESTS

Add an option to reset gcov counters before running tests.
This ensures that only code lines triggered by test itself are counted in
the coverage report and all the board initialization code and pre-test
bootstrap is not counted. This is useful when, for example, you are
testing code that is also executed during bootup

Test Plan:
west build -p always -b qemu_x86_64 tests/ztest/base/ -- \
-DCONFIG_COVERAGE=y -DCONFIG_COVERAGE_GCOV=y -DCONFIG_COVERAGE_DUMP=y \
-DCONFIG_ZTEST_COVERAGE_RESET_BEFORE_TESTS=y
ninja -Cbuild run | tee log.log

Signed-off-by: Roman Studenikin <srv@meta.com>

srv-meta commented:
> would be great to see this used in the tree, i.e. add an kconfig option that would allow us to reset counters before running a test.

@nashif:  added a Kconfig option

ycsin commented on subsys/testsuite/coverage/coverage.c:
```diff
@@ -232,6 +232,48 @@ size_t gcov_to_gcda(uint8_t *buffer, struct gcov_info *info)
+	}
+}
+
+void gcov_reset_all_counts() {
```
`checkpatch` can help to visualize this kind of issue from the vscode editor itself.

https://github.com/ycsin/zephyr/blob/ba9cc24441b50d9cde5254bde52b76748e119e2d/.devcontainer/devcontainer.json#L10-L43

ycsin reviewed (approved):
Refresh +1

ycsin commented:
ping @nashif @jgl-meta @cfriedt @nordic-piks could you please take another look? Thanks

jgl-meta reviewed (approved):
lgtm
